js/app/modules/filter/currentSet.js

```diff
@@ -51,25 +51,32 @@ var CurrentSet = new Module.Class({
         return belongs_to_set && !belongs_to_subset;
     },
 
-    // Filter override
-    // This Filter can modify the Xapian query, but it can't express its whole
-    // filtering logic in the Xapian query, so we still need to request more
-    // models in case some returned models must be dropped.
-    can_modify_xapian_query: function () {
-        return false;
-    },
-
     // Filter implementation
     modify_xapian_query_impl: function (query) {
+        let tags = this._current_set.child_tags;
+
+        if (this.invert) {
+            if (query.excluded_tags.indexOf(tags[0]) !== -1)
+                return query;
+        } else {
+            if (query.tags_match_any.indexOf(tags[0]) !== -1)
+                return query;
+        }
+
+        let subtags = [];
```

Comment:
`concat()` in a loop is creating a lot of new array objects and discarding them. Better would be `const subtags = SetMap.get_children_sets(this._current_set).map(set => { ... });`